### PR TITLE
feat(talk): Add reactions, replies, polls, and file sharing tools

### PR DESCRIPTION
**Pull request by Pavlinchen**
## Summary

- Extends Talk from 4 to 15 MCP tools, adding reactions, reply-to-message, polls, and file sharing
- Enhances `list_talk_conversations` and `list_messages_in_conversation` to return structured data with tokens and message IDs
- All new tools use `nc.ocs()` for OCS API calls, consistent with `shares.py`

## Motivation

The current Talk module only supports listing conversations, reading messages, sending messages, and creating public rooms. This is insufficient for meaningful AI assistant interaction — the assistant cannot react to messages, reply in threads, create polls for decision-making, or share files in conversations. These are core Talk features that users expect the AI to interact with.

### Disclaimer

Implementation from a technical perspective was done mainly by a LLM (Claude Opus 4.7 max effort). 
Every tool and endpoint response shape was validated end-to-end against Nextcloud 32.0.8 + Talk 22.0.9, with a full lifecycle smoke test completed before opening the PR.

### Architecture

All new tools follow the existing pattern of accepting `conversation_name` (string) and resolving to a token internally via a shared `_get_token()` helper. This maintains backward compatibility and consistency with the existing name-based tool interfaces.

OCS endpoints used:
- Reactions: `/ocs/v2.php/apps/spreed/api/v1/reaction/{token}/{messageId}`
- Chat (replies): `/ocs/v2.php/apps/spreed/api/v1/chat/{token}`
- Polls: `/ocs/v2.php/apps/spreed/api/v1/poll/{token}`
- File sharing (send): `/ocs/v2.php/apps/files_sharing/api/v1/shares` (core Share API, shareType=10)
- File sharing (list): `/ocs/v2.php/apps/spreed/api/v1/chat/{token}/share`

### Enhanced existing tools

| Tool | Before | After |
|------|--------|-------|
| `list_talk_conversations` | `["name1", "name2"]` | `[{name, token, type, unread_messages}]` |
| `list_messages_in_conversation` | `["timestamp actor: text"]` | `[{id, timestamp, actor, message, is_replyable, reactions}]` |

The enhanced return formats are necessary — without message IDs the LLM cannot target reactions or replies, and without tokens the LLM cannot use the OCS endpoints.

### New tools

| Tool | Decorator | Feature | Required Capability |
|------|-----------|---------|-------------------|
| `add_reaction` | dangerous | Reactions | `reactions` |
| `remove_reaction` | dangerous | Reactions | `reactions` |
| `list_reactions` | safe | Reactions | `reactions` |
| `reply_to_message` | dangerous | Replies | `chat-replies` |
| `create_poll` | dangerous | Polls | `talk-polls` |
| `get_poll` | safe | Polls | `talk-polls` |
| `vote_on_poll` | dangerous | Polls | `talk-polls` |
| `close_poll` | dangerous | Polls | `talk-polls` |
| `share_file_to_conversation` | dangerous | Files | (core) |
| `list_shared_files` | safe | Files | `rich-object-list-media` |
| `list_shared_items_overview` | safe | Files | `rich-object-list-media` |

### Unchanged tools

- `create_public_conversation` — unchanged
- `send_message_to_conversation` — unchanged (uses nc_py_api wrapper)

## Test Plan

Tested all endpoints against Nextcloud 32.0.8 + Talk 22.0.11:

- [x] **Reactions**: add (201), list (200 with emoji-keyed actors), remove (200, verified empty)
- [x] **Reply**: send with replyTo (201, response includes parent.id reference)
- [x] **Polls**: create (201), get (200, status=open), vote (200, votedSelf populated), close (200, final votes visible)
- [x] **File sharing**: share file to conversation via shareType=10 (200), list shared files (200, includes file metadata), overview (200, items grouped by type)
- [x] **Enhanced tools**: list_talk_conversations returns tokens, list_messages returns message IDs + is_replyable + reactions
- [x] Integration test with Context Agent MCP server (requires running ExApp)

## Required Talk Capabilities

The following Talk capabilities are used by the new tools (all available since Talk 15+):
- `reactions` — for reaction tools
- `chat-replies` — for reply_to_message
- `talk-polls` — for poll tools
- `rich-object-list-media` — for file listing tools
- `silent-send` — for the silent parameter on reply_to_message
- `media-caption` — for the caption parameter on share_file_to_conversation

### CODE REVIEW

File: ex_app/lib/all_tools/talk.py [hunk 1]
```diff
@@ -8,64 +10,281 @@
...
+		})
+
+	@tool
+	@safe_tool
+	async def get_poll(conversation_name: str, poll_id: int):
+		"""
+		Get the current state and results of a poll
+		:param conversation_name: The name of the conversation (obtainable via list_talk_conversations)
+		:param poll_id: The id of the poll (obtainable from create_poll or from message parameters in list_messages_in_conversation)
+		:return: poll data including question, options, votes, status, and who voted
```
Comment (marcelklehr) [inline ex_app/lib/all_tools/talk.py]: It could make sense to add a note here that options are 0-indexed as the llm in my test was off by one, labelling the votes.

File: ex_app/lib/all_tools/talk.py [hunk 2]
```diff
@@ -8,64 +10,281 @@
...
+	async def list_shared_files(conversation_name: str, limit: int = 100):
+		"""
+		List files that have been shared in a Talk conversation
+		:param conversation_name: The name of the conversation (obtainable via list_talk_conversations)
+		:param limit: Maximum number of results (default 100, max 200)
+		:return: list of chat messages containing shared files with file metadata (name, size, mimetype, link)
+		"""
+		token = await _get_token(conversation_name)
+		return await nc.ocs('GET', f'/ocs/v2.php/apps/spreed/api/v1/chat/{token}/share', params={
+			'objectType': 'file',
```
Comment (marcelklehr) [inline ex_app/lib/all_tools/talk.py]: Should we parameterize this to be able to list other types of shared items? (audio, media, poll etc)

**Comment by marcelklehr** [comment]
Tested and confirmed working

File: ex_app/lib/all_tools/talk.py (hunk 1 above)
Comment (Pavlinchen) [reply ex_app/lib/all_tools/talk.py]: Thanks for pointing that out, that slipped through in my review.
I've added a note inspired by the docstring from `vote_on_poll` via commit [5b0aa9c](https://github.com/nextcloud/context_agent/pull/157/commits/5b0aa9c)

What do you think?

File: ex_app/lib/all_tools/talk.py (hunk 2 above)
Comment (Pavlinchen) [reply ex_app/lib/all_tools/talk.py]: While that was not in my original scope, I think it makes sense to do that while we're at it. 
Please give me a bit of time to do a bit of thinking and testing and I'll get back to you on that.

File: ex_app/lib/all_tools/talk.py (hunk 1 above)
Comment (marcelklehr) [reply ex_app/lib/all_tools/talk.py]: Looks good!

File: ex_app/lib/all_tools/talk.py (hunk 2 above)
Comment (Pavlinchen) [reply ex_app/lib/all_tools/talk.py]: Thanks for your patience.
Added now and successfully tested on my side. 
Also wrapped everything as strings, as you requested here https://github.com/nextcloud/context_agent/pull/155#discussion_r3130722515

**Comment by Pavlinchen** [comment]
Hello @marcelklehr friendly reminder on this PR

**Comment by Pavlinchen** [comment]
(force pushed to use alias e-Mail adress)

**Review by marcelklehr — APPROVED** [review]

**Comment by Pavlinchen** [comment]
Thank you for the fast actions!
I can see you added the "needs documentation" label, is that related to https://docs.nextcloud.com/server/latest/admin_manual/ai/app_context_agent.html ?
If so, is it intended for me to make the updates at https://github.com/nextcloud/documentation, or is that an internal thing?

**Comment by marcelklehr** [comment]
> is it intended for me to make the updates at [nextcloud/documentation](https://github.com/nextcloud/documentation), or is that an internal thing?

I added the label mostly for myself, but if you'd like to lend a hand, feel free to submit a PR for [docs.nextcloud.com/server/latest/admin_manual/ai/app_context_agent.html](https://docs.nextcloud.com/server/latest/admin_manual/ai/app_context_agent.html) :blue_heart:

**Comment by Pavlinchen** [comment]
Seems doable, I will also create a PR for the new Deck tools while I'm at it.

**Comment by Pavlinchen** [comment]
done

**Comment by marcelklehr** [comment]
@Pavlinchen Can you cc me on these? thanks

**Comment by Pavlinchen** [comment]
sorry, what exactly do you mean by "cc"? Just pinging you in the comments? 
I've mentioned the "causing" PR in nextcloud/context_agent in each nextcloud/docs PR, so you can also see them in the PR activity (^)

**Comment by marcelklehr** [comment]
ah, thanks :)

**Comment by Pavlinchen** [comment]
😉
